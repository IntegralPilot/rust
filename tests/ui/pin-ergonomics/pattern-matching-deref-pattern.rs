//@ revisions: pin_ergonomics normal
//@ edition:2024
//@ check-pass
// //@[normal] check-pass
// //@[pin_ergonomics] rustc-env:RUSTC_LOG=rustc_hir_typeck::expr_use_visitor=DEBUG
#![cfg_attr(pin_ergonomics, feature(pin_ergonomics))]
#![feature(deref_patterns)]
#![allow(incomplete_features)]

// This test verifies that the `pin_ergonomics` feature works well
// together with the `deref_patterns` feature.

use std::pin::Pin;

struct Foo<T, U> {
    x: T,
    y: U,
}

struct Bar<T, U>(T, U);

enum Baz<T, U> {
    Foo(T, U),
    Bar { x: T, y: U },
}

fn foo_mut<T: Unpin, U: Unpin>(foo: Pin<&mut Foo<T, U>>) {
    let Foo { .. } = foo;
    let Pin { .. } = foo;
    let _ = || {
        let Foo { .. } = foo;
        let Pin { .. } = foo;
    };

    #[cfg(pin_ergonomics)]
    let Foo { x, y } = foo;
    #[cfg(pin_ergonomics)]
    let _ = || {
        let Foo { x, y } = foo;
    };
}

fn foo_const<T: Unpin, U: Unpin>(foo: Pin<&Foo<T, U>>) {
    let Foo { .. } = foo;
    let Pin { .. } = foo;
    let _ = || {
        let Foo { .. } = foo;
        let Pin { .. } = foo;
    };

    #[cfg(pin_ergonomics)]
    let Foo { x, y } = foo;
    #[cfg(pin_ergonomics)]
    let _ = || {
        let Foo { x, y } = foo;
    };
}

fn bar_mut<T: Unpin, U: Unpin>(bar: Pin<&mut Bar<T, U>>) {
    let Bar(..) = bar;
    let Pin { .. } = bar;
    let _ = || {
        let Bar(..) = bar;
        let Pin { .. } = bar;
    };

    #[cfg(pin_ergonomics)]
    let Bar(x, y) = bar;
    #[cfg(pin_ergonomics)]
    let _ = || {
        let Bar(x, y) = bar;
    };
}

fn bar_const<T: Unpin, U: Unpin>(bar: Pin<&Bar<T, U>>) {
    let Bar(..) = bar;
    let Pin { .. } = bar;
    let _ = || {
        let Bar(..) = bar;
        let Pin { .. } = bar;
    };

    #[cfg(pin_ergonomics)]
    let Bar(x, y) = bar;
    #[cfg(pin_ergonomics)]
    let _ = || {
        let Bar(x, y) = bar;
    };
}

fn main() {}
