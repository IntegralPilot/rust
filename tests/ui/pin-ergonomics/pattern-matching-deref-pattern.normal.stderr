error[E0507]: cannot move out of a shared reference
  --> $DIR/pattern-matching-deref-pattern.rs:45:28
   |
LL |         let Foo { x, y } = foo.as_mut();
   |                   -  -     ^^^^^^^^^^^^
   |                   |  |
   |                   |  ...and here
   |                   data moved here
   |
   = note: move occurs because these variables have types that don't implement the `Copy` trait
help: consider borrowing the pattern binding
   |
LL |         let Foo { ref x, y } = foo.as_mut();
   |                   +++
help: consider borrowing the pattern binding
   |
LL |         let Foo { x, ref y } = foo.as_mut();
   |                      +++

error[E0507]: cannot move out of a shared reference
  --> $DIR/pattern-matching-deref-pattern.rs:35:24
   |
LL |     let Foo { x, y } = foo.as_mut();
   |               -  -     ^^^^^^^^^^^^
   |               |  |
   |               |  ...and here
   |               data moved here
   |
   = note: move occurs because these variables have types that don't implement the `Copy` trait
help: consider borrowing the pattern binding
   |
LL |     let Foo { ref x, y } = foo.as_mut();
   |               +++
help: consider borrowing the pattern binding
   |
LL |     let Foo { x, ref y } = foo.as_mut();
   |                  +++

error[E0507]: cannot move out of a shared reference
  --> $DIR/pattern-matching-deref-pattern.rs:67:28
   |
LL |         let Foo { x, y } = foo.as_ref();
   |                   -  -     ^^^^^^^^^^^^
   |                   |  |
   |                   |  ...and here
   |                   data moved here
   |
   = note: move occurs because these variables have types that don't implement the `Copy` trait
help: consider borrowing the pattern binding
   |
LL |         let Foo { ref x, y } = foo.as_ref();
   |                   +++
help: consider borrowing the pattern binding
   |
LL |         let Foo { x, ref y } = foo.as_ref();
   |                      +++

error[E0507]: cannot move out of a shared reference
  --> $DIR/pattern-matching-deref-pattern.rs:57:24
   |
LL |     let Foo { x, y } = foo.as_ref();
   |               -  -     ^^^^^^^^^^^^
   |               |  |
   |               |  ...and here
   |               data moved here
   |
   = note: move occurs because these variables have types that don't implement the `Copy` trait
help: consider borrowing the pattern binding
   |
LL |     let Foo { ref x, y } = foo.as_ref();
   |               +++
help: consider borrowing the pattern binding
   |
LL |     let Foo { x, ref y } = foo.as_ref();
   |                  +++

error[E0507]: cannot move out of a shared reference
  --> $DIR/pattern-matching-deref-pattern.rs:89:25
   |
LL |         let Bar(x, y) = bar.as_mut();
   |                 -  -    ^^^^^^^^^^^^
   |                 |  |
   |                 |  ...and here
   |                 data moved here
   |
   = note: move occurs because these variables have types that don't implement the `Copy` trait
help: consider borrowing the pattern binding
   |
LL |         let Bar(ref x, y) = bar.as_mut();
   |                 +++
help: consider borrowing the pattern binding
   |
LL |         let Bar(x, ref y) = bar.as_mut();
   |                    +++

error[E0507]: cannot move out of a shared reference
  --> $DIR/pattern-matching-deref-pattern.rs:79:21
   |
LL |     let Bar(x, y) = bar.as_mut();
   |             -  -    ^^^^^^^^^^^^
   |             |  |
   |             |  ...and here
   |             data moved here
   |
   = note: move occurs because these variables have types that don't implement the `Copy` trait
help: consider borrowing the pattern binding
   |
LL |     let Bar(ref x, y) = bar.as_mut();
   |             +++
help: consider borrowing the pattern binding
   |
LL |     let Bar(x, ref y) = bar.as_mut();
   |                +++

error[E0507]: cannot move out of a shared reference
  --> $DIR/pattern-matching-deref-pattern.rs:111:25
   |
LL |         let Bar(x, y) = bar.as_ref();
   |                 -  -    ^^^^^^^^^^^^
   |                 |  |
   |                 |  ...and here
   |                 data moved here
   |
   = note: move occurs because these variables have types that don't implement the `Copy` trait
help: consider borrowing the pattern binding
   |
LL |         let Bar(ref x, y) = bar.as_ref();
   |                 +++
help: consider borrowing the pattern binding
   |
LL |         let Bar(x, ref y) = bar.as_ref();
   |                    +++

error[E0507]: cannot move out of a shared reference
  --> $DIR/pattern-matching-deref-pattern.rs:101:21
   |
LL |     let Bar(x, y) = bar.as_ref();
   |             -  -    ^^^^^^^^^^^^
   |             |  |
   |             |  ...and here
   |             data moved here
   |
   = note: move occurs because these variables have types that don't implement the `Copy` trait
help: consider borrowing the pattern binding
   |
LL |     let Bar(ref x, y) = bar.as_ref();
   |             +++
help: consider borrowing the pattern binding
   |
LL |     let Bar(x, ref y) = bar.as_ref();
   |                +++

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0507`.
